awx_collection/plugins/modules/tower_application.py

```diff
@@ -0,0 +1,170 @@
+      type: str
+      required: True
+    organization:
+      description:
+        - Name of organization for application.
+      type: str
+      required: True
+    redirect_uris:
+      description:
+        - Allowed urls list, space separated. Required when authorization-grant-type=authorization-code
+      default: "present"
+      choices: ["present", "absent"]
+      type: str
+    state:
+      description:
+        - Desired state of the resource.
+      default: "present"
+      choices: ["present", "absent"]
+      type: str
+    skip_authorization:
+      description:
+        - Set True to skip authorization step for completely trusted applications.
+      default: false
+      type: bool
+    username:
+      description:
+        - Username for this credential. ``access_key`` for AWS.
+        - Deprecated, please use inputs
+      type: str
+'''
+
+
+EXAMPLES = '''
+- name: Add Foo application
+  tower_application:
+    name: "Foo"
+    description: "Foo bar application"
+    organization: "test"
+    state: present
+    authorization-grant-type: password
+    client-type: public
+
+- name: Add Foo application
+  tower_application:
+    name: "Foo"
+    description: "Foo bar application"
+    organization: "test"
+    state: present
+    authorization-grant-type: authorization-code
+    client-type: confidential
+    redirect_uris: http://tower.com/api/v2/
+'''
+
+import time
+
+from ..module_utils.tower_api import TowerModule
+
+
+def main():
+    # Any additional arguments that are not fields of the item can be added here
+    argument_spec = dict(
+        name=dict(required=True),
+        description=dict(),
+        authorization_grant_type=dict(required=True),
+        client_type=dict(required=True, choices=['public', 'confidential']),
+        organization=dict(required=True),
+        redirect_uris=dict(type="list", elements='str'),
+        state=dict(choices=['present', 'absent'], default='present'),
+        skip_authorization=dict(type=bool),
+        username=dict()
+    )
+
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec)
+
+    # Extract our parameters
+    name = module.params.get('name')
+    description = module.params.get('description')
+    authorization_grant_type = module.params.get('authorization_grant_type')
+    client_type = module.params.get('client_type')
+    organization = module.params.get('organization')
+    redirect_uris = ' '.join(module.params.get('redirect_uris'))
+    state = module.params.get('state')
+
+    # Attempt to look up the related items the user specified (these will fail the module if not found)
+    org_id = module.resolve_name_to_id('organizations', organization)
+
+    # Attempt to look up application based on the provided name and org ID
+    application = module.get_one('applications', **{
+        'data': {
+            'name': name,
+            'organization': org_id
+        }
+    })
```

Comment:
FYI, this will make it so that you can't pass in an ID for an application and have it picked up. This is a general problem we have with the collection and will need to address.

Reply:
I am not sure to understand. Does this need an edit ?

Reply:
No, you don't have to address this at this time. We will have to come up with a global fix for the collection.